packages/matter.js/src/codec/BtpCodec.ts

```diff
@@ -0,0 +1,181 @@
+    RESERVE_BIT = 0b00000000,
+}
+
+export const enum ProtocolVersion {
+    FINAL_PROTOCOL_VERSION = 0b00000100,
+}
+
+export class BtpCodec {
+
+    static decodeHandshakeRequest(data: ByteArray): HandshakeRequest {
+        const reader = new DataReader(data, Endian.Little);
+
+        return {
+            header: this.decodeHeader(reader),
+            payload: this.decodeRequestPayload(reader)
+        };
+
+    }
+
+    static decodePacket(data: ByteArray): Packet {
+        const reader = new DataReader(data, Endian.Little);
+
+        return {
+            header: this.decodeHeader(reader),
+            payload: this.decodeRacketPayload(reader)
+        };
+
+    }
+
+    static encodeHandshakeResponse({ header, payload }: HandshakeResponse): ByteArray {
+        
+        return ByteArray.concat(
+            this.encodeHeader(header),
+            this.encodeResponsePayload(payload)
+        );
+    }
+
+    private static decodeRacketPayload(reader: DataReader<Endian.Little>): PacketPayload {
+
+        const opcode = reader.readUInt8();
+        const isManagementOpcode = (opcode & Opcode.MANAGEMENT_OPCODE) !== 0;
+
+        if (opcode !== Opcode.MANAGEMENT_OPCODE) throw new Error("Handshake Error - Management Opcode is incorrect");
+
+        const ackNumber = reader.readUInt8();
+        const sequenceNumber = reader.readInt8();
+        const msgLength = reader.readInt16();
+        const segmentPayload = reader.getRemainingBytes();
+
+        return { isManagementOpcode, ackNumber, sequenceNumber, msgLength, segmentPayload };
+    }
+
+
+    private static decodeRequestPayload(reader: DataReader<Endian.Little>): HandshakeRequestPayload {
+
+        const opcode = reader.readUInt8();
+        const isManagementOpcode = (opcode & Opcode.MANAGEMENT_OPCODE) !== 0;
+
+        if (opcode !== Opcode.MANAGEMENT_OPCODE) throw new Error("Handshake Error - Management Opcode is incorrect");
+
+        const version = reader.readUInt32();
+        const attMtu = reader.readInt16();
+        const clientWindowSize = reader.readInt8();
+
+        return { isManagementOpcode, version, attMtu, clientWindowSize };
+    }
+
+    private static decodeHeader(reader: DataReader<Endian.Little>): Header {
+
+        const headerBits = reader.readUInt8();
+        const isUnused1 = (headerBits & HeaderBits.UNUSED_BIT) == 0;
+        const isHandshake = (headerBits & HeaderBits.HANDSHAKE_BIT) !== 0;
+        const isManagement = (headerBits & HeaderBits.MANAGEMENT_MESSAGE) !== 0;
+        const isUnused2 = (headerBits & HeaderBits.UNUSED_BIT) == 0;
+        const isAckMsg = (headerBits & HeaderBits.ACKNOWLEDGEMENT_MSG) == 0;
+        const isEndingSegment = (headerBits & HeaderBits.ENDING_SEGMENT) !== 0;
+        const isUnused3 = (headerBits & HeaderBits.UNUSED_BIT) == 0;
+        const isBeginingSegment = (headerBits & HeaderBits.BEGINNING_SEGMENT) !== 0;
+
+        if (headerBits !== HEADER) throw new Error("Error in BTP Handhshake Request Headers");
```

Comment:
But this is wrong when used for a "non handshake package"

Reply:
Updated, please review